### PR TITLE
Recursive type equality check for de-duplicating

### LINKED ISSUE #19: Type deduplication and ensuring unique type paths
This issue explains the background of a problem that we currently have about generating  rust data types from a `PortableRegistry`.

### Deduplication

A `PortableRegistry` is part of a substrate node's metadata and contains all the types necessary to interact with the node. Only **concrete types** can be registered into a `PortableRegistry`. Each type registered gets its own type id. When registering a struct, the types of all its field are registered recursively as well. Imagine we register a struct `Bar` into the registry:
```
struct Bar {
    a: Foo<u8>,
    b: Foo<u32>
}

struct Foo<T>{
    t: T
}
```
The types `Foo<u8>` and `Foo<u32>` will be distinct types in the registry with their own type ids. But: they can be represented with just one rust type: `Foo<T>`.  So they are **duplicates**. When generating rust code, we want to **deduplicate** them and just generate one `Foo<T>` struct. We want to deduplicate all types that have the **same type path** and **same structure**.
Comparing the type path (e.g. `my::module::Foo`) is easy, comparing the structure not so much, more on that later.

### Ensuring unique type paths

Another thing we can encounter is that the type registry contains two types that have the **same type path** but a **different structure**. In these cases we *cannot* generate a single rust type. Instead we can give each of the two types a distinct name (changing the ident of the type path) and generate distincts rust types. E.g. `Foo -> Foo1 and Foo2`. 

There are 3 cases when this happens AFAIK:
1. feature flags that change the structure of a type. In this case the type path will still be the same.
2. manual implementation of the `scale_info::TypeInfo` trait, specifying the same type path for two types.
3. associated types: Imagine you have a type like this:
```
trait X{
     type Assoc;
}
struct A;  
impl X for A {  type Accoc = u8;  }
struct B;  
impl X for B {  type Accoc = u32;  }

#[scale_info(skip_type_params(T))] // this is optional
struct Foo<T: X>{
    field: X::Assoc,
}
```
Then if `Foo<A>` and `Foo<B>` get registered, there is no way for us to generate a single rust type `Foo<T>` anymore. The registry does not contain the trait bounds, and we cannot generic types with trait bounds and associated types as fields.
So because there is simply no way to generate struct `Foo<T: X>{ field: X::Assoc }` we generate `Foo1 { field: u8 }` and `Foo2 { field: u32 }` instead.

**But:** We do not want to tamper with the users `PortableRegistry` by default. So if a case like this is encountered we currently return an `Error`. The user can run the `scale_typegen::utils::ensure_unique_type_paths` fn on their `PortableRegistry` to remap duplicate type paths `Foo -> Foo1 and Foo2` where the type structure differs. This makes the type generation succeed.

## What is the problem?

The issue is that we are currently not very good at "deduplicating" types because it is not so easy to determine if the structure of two types is the same considering their type params. The logic for this is in the `scale_typegen::utils::types_equal_extended_to_params` function.

To see if two types A and B are equal all of their fields need to equal: For each field we compare the type ids. If they are the same, we consider the types equal. If they are not the same, but they are the type id of a type param for A and B, and this type params has the same name, then we also consider them equal. 

There are a few problems with that: Imagine we have these two types:
```
struct Foo<T = u64> {
    inner: Vec<u64>
}
struct Foo<T = u32> {
    inner: Vec<u32>
}
```
There are now 6 types in the type registry, let's give them arbitrary ids:
```
Foo<T=u64>: 0
Vec<u64>: 1
u64: 2
Foo<T=u32>: 3
Vec<u32>: 4
u32: 5
```

Then when comparing the `inner` fields, their ids are  `1 (Vec<u64>)` and `4 (Vec<u32>)` respectively, but the generic param T has ids `2 (u64)` for `Foo<T=u64>` and `5 (u32)` for `Foo<T=u32>`, so they are deemed to be different. 
This means we cannot detect that Foo<u32> and Foo<u64> could be one rust struct. We deduplicate them into Foo1<u32> and Foo2<u64> or the codegen will return an error.

Solution: We need to recursively go down into the Vec<u64> and Vec<u32> types as well. Just calling `types_equal_extended_to_params` on them though is not enough, we need to pass down more information about type_id -> type param mapping from the parent struct. 
Here we also run into the issue that only some fields with said type id might map to the generic and others are actually fixed:
Consider `Foo<T=u32>{ a: T, b:u32 }`. 

Another case where we currently return false, is when comparing these two types, suppose they have the same type path:
```
struct Foo<T> {
    inner: T
}
struct Foo<B> {
    inner: B
}
```
This is because the `type_name` of the field `inner` does not match even though structurally they are the same. 

## A hashing solution?

One solution I could think of is that we do not compare two types directly. Instead we calculate a `hash` for each type and insert them into a bin for each `hash`. All types that land in the same `bin` can be considered equal and we can pick any one of them to generate the rust type.

This hashing just needs to be aware of type params, and hash the same type params in the same way. 
Here the hash should be based on the position of the type params in the type definition and not on their name.
First: build up a table of params and their position: 
`[(A, 0), (B,1), (C,2), ...]`
When iterating over the fields we can check for each field if it maps to a type param or not:
- if type's `type_name` and `type_id` match with a type param: -> `hasher.hash(position of type param); hasher.hash(GENERIC_FLAG)`
- else if type's `type_name` contains angle brackets e.g. `Vec<B>`:  calculate hash of `Vec<B>`:
    - here, we pass [(B, 1)] to `Vec<B>` and use it instead of `[(B, 0)]` that we would get by just looking at `Vec<B>` in a vacuum. 
    - when recursion down done, `hasher.hash(hash of Vec<B>)`.
- else: type does not refer to type param: -> `hasher.hash(type id); hasher.hash(NOT_GENERIC_FLAG);`

### PR DESCRIPTION
**Pull request by jsdw**
Recurse through types to check for equality.

The basic algorithm here is:
- If type IDs match, they are the same.
- If type IDs can be explained by the same generic parameter, they are the same.
- If type paths or generic names don't match, they are different.
- If the corresponding TypeDefs (shape of type) is different, they are different.
- Else, recurse through any contained type IDs and start from the top.

Unfortunately, we aren't strictly improving on our type generation from before, ie there are some types that would previously be deduplicated but now arent. Consider these types:

```txt
1: struct Foo<T = u8, U = u8, V = u8>(u8, u8, u8);
2: struct Foo<T = u8, U = u16, V = u32>(u8, u16, u32);
3: struct Foo<T = u8, U = u16, V = u32>(u32, u16, u8);
```

Under the previous approach, we'd gather a set of generics together for each type and see if it could explain both types. However, with this approach, 1 == 2 and 1 == 3 (both could be represented by `Foo<T,U,V>(T,U,V)`) but 2 != 3.
So in extreme cases, it could de-duplicate types that are actually different.

Under the new approach, we map each type to exactly 1 generic param, and fail if they don't both map to the same one. This will miss certain cases, but hopefully also can't incorrectly de-duplicate things. Maybe we can improve this, but I suspect that owuld involve carrying around the actualy generic param mappings with the types so that we can avoid deduplicating where they differ. Needs thought.

### Testing done

- Subxt seems to end up with OK metadata offhand; the new code shaved about 1k lines off a 72k line file (so not too many) and comparing a couple of the differences, everything looked sensible. 
- Specifically, only one BoundedVec and similar now!
- The Subxt tests also passed, and I added a local test, but in the end it's hard to spot edge cases with this stuff!

Closes #19

### CODE REVIEW

File: typegen/src/utils.rs [hunk 1]
```diff
@@ -76,110 +73,262 @@ pub fn ensure_unique_type_paths(types: &mut PortableRegistry) {
...
+        a_params: &GenericsList,
+        b: &[Field<PortableForm>],
+        b_params: &GenericsList
+    | -> bool {
         if a.len() != b.len() {
             return false;
         }
         a.iter().zip(b.iter()).all(|(a, b)| {
-            a.name == b.name && a.type_name == b.type_name && ids_equal(a.ty.id, b.ty.id)
+            a.name == b.name
```
Comment (pkhry) [inline typegen/src/utils.rs:186]:
```suggestion
        let equal_name = a.name == b.name;

        if let (Some(a_type_name), Some(b_type_name)) = (&a.type_name, &b.type_name) {
            let type_names_present = a_params.index_for_type_id(a.ty.id).is_some()
                && b_params.index_for_type_id(b.ty.id).is_some();
            equal_name
                && a_type_name == b_type_name
                && !a_type_name.contains("::")
                && type_names_present
        } else {
            if a_params.index_for_type_id(a.ty.id) == b_params.index_for_type_id(b.ty.id)
                || a.type_name.as_ref().is_some_and(|x| x.contains("::"))
            {
                equal_name && types_equal_recurse(a.ty.id, a_params, b.ty.id, b_params)
            } else {
                dbg!(a, b);
                false
            }
        }
```

Comment (pkhry) [reply typegen/src/utils.rs:186]: this should not break any things, but the tests that were commented out should work

Comment (jsdw) [reply typegen/src/utils.rs:186]: I don't really follow the logic above offhand! Has our element conversation resolved this comment for you, or do you still think there's some improvement that we can make above?

**Review by pkhry — APPROVED** [review]
LGTM

**Review by pkhry — APPROVED** [review]
LGTM

**Review by niklasad1 — APPROVED** [review]
